src/store_client.cc

```diff
@@ -496,11 +517,32 @@ store_client::fileRead()
               this);
 }
 
+bool
+store_client::parseHttpHeader(const ssize_t len)
+{
+    startBufferingReplyBytes();
+    replyHeaderBytes->append(copyInto.data, len);
```

Comment:
> Now, we parsed the 6KB response header and want to return the requested data to the store_client user

We do not do this here. _replyHeaderBytes_ contains both header and body start bytes. As soon as we managed to parse headers, we return the entire buffer to the caller, which writes all these bytes to Store (with a single mem_obj->write() call) - we do not care about  copyInto leftovers because they have been copied to _replyHeaderBytes_.

Reply:
>  I see no code that would give the user something other than copyInto.data 

When we send headers to the client, we get them from the HttpReply object, and  copyInto.data is ignored. When we have written reply to the client, copyInto.offset is updated to the headers offset (in our case, say ~6KB). When we come again for the next data block, we read from the Store starting from this headers offset, and copyInto gets the initial body bytes (i.e., 4KB). In other words, your we read 4KB+4KB, then send 6KB header to the client (discarding the remaining 2KB), and then read 4KB body (including these initial 2KB, which are read twice).

I performed several tests with different header/body sizes - they were OK.

Reply:
> ... initial 2KB, which are read twice

Aha, so we may be reading the same response body prefix twice from Store. I missed that possibility, and it explains how we can get the 2KB post-header data to the client on the second client request (i.e. we read it again). Does unpatched Squid also go to disk the second time for the already read data if store_client reads more than just the response header on the first read? Or is this a new performance problem that is limited to patched Squid dealing with larger-than-buffer response headers?

> I performed several tests with different header/body sizes - they were OK.

```C++
    char *buf = next()->readBuffer.data;
    char *body_buf = buf + reply->hdr_sz; // XXX: Now body_buf points 2KB past readBuffer.data end!
```

I understand that the above and similar code can still miraculously work because there are ten other adjustments of various offsets/sizes/pointers there, and there is a chance it all just comes together despite bogus pointers, negative offsets, etc. in the middle of it. While we should not unleash this insanity on ourselves, my primary concern right now is not whether the new code "works", but how the tests _exercise_ this particular path/case. It would be great to be able to rely on tests as we adjust this code!

Please let me know (here or on Jira) what test to run on my side to see how much of Squid code is exposed to these bogus buffers and related offsets.

Reply:
> Does unpatched Squid also go to disk the second time for the already read data if store_client reads more than just the response header on the first read?

No, it does not. The body bytes, following the header, are appended in to the result MemBuf Stream::sendStartOfMessage(). To be precise, the patched Squid has these body bytes already in memory - so it goes there the second time.

>  the above and similar code can still miraculously work

Agreed. In your example, _body_buf_ is set to point outside the buf-allocated memory (4KB) (in my test with 6K header). The miracle happens probably because this pointer is never dereferenced, since _body_size_ is set to zero above (in a calculation with _reqofs_).

> what test to run on my side to see how 

I did not create any special test for this, but [added](https://bitbucket.org/measurementfactory/daft/branch/SQUID-479-large-header) a configurable (large) header to our Daft cache-response test.

Reply:
The initial discussion here focused on HTTP/client_side_reply.cc (e.g., SendMoreData()) handling of Store data returned by store_client. There are other storeClientCopy() users, of course:

* asHandleReply(): Expects WHOIS response body data where the HTTP headers[^1] are! Unaware of hdr_sz.
* peerDigestHandleReply(): Parses received HTTP headers in peerDigestSwapInHeaders()? Unaware of hdr_sz.
* urnHandleReply(): Uses a single fixed-size buffer to accumulate multiple Store reads. Parses received HTTP headers. Unaware of hdr_sz.
* netdbExchangeHandleReply(): Parses received HTTP headers. Unaware of hdr_sz (computes its replacement).
* Mgr::StoreToCommWriter::noteStoreCopied(): Blindly writes received raw buffer to the client socket. Unaware of hdr_sz.

... where "hdr_sz" above is the `mem_obj->reply->hdr_sz` that helps client_side_reply to skip header suffix in the copyInto buffer.

AFAICT, leaving header suffix in the first copyInto buffer returned to the client will break all of the above storeClientCopy() users. Some of them may never get a large response header, but none of them are guaranteed to never get it by their own design -- their fate is in somebody else hands, and that somebody does not know about it. Granted, most of the above callers are already broken for other reasons, but we would be making their situation significantly worse if we start feeding them HTTP response header suffixes.

No code changes requested. I am primarily writing this summary of storeClientCopy() users to assist us in selecting the best long-term solution...

[^1]: Generated by our WhoisState::setReplyToOK() code.